File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        
        int n = diffs.length;
        
        int[] diff_arr = new int[n + 1];
        int[] time_arr = new int[n + 1];
        
        for(int i = 0; i < n; i++){
            diff_arr[i + 1] = diffs[i];
            time_arr[i + 1] = times[i];
        }    
        
        int left = 0;
        int right = Integer.MAX_VALUE;
        int ans = 0;
        
        while(left <= right) {
            
            int mid = (left + right) / 2;
            
            long result = solve(mid, diff_arr, time_arr);
            
            if(result <= limit){
                ans = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return Math.max(1, ans);
    }
    
    public long solve(int level, int[] diff_arr, int[] time_arr) {

        long ret = 0;
        
        for(int i = 1; i < diff_arr.length; i++) {
            
            if(diff_arr[i] <= level) {
                ret += time_arr[i];
            }
            else {
                int diff = diff_arr[i] - level;
                ret += time_arr[i - 1] * diff;
                ret += time_arr[i] * (diff + 1);
            }
        }
        
        return ret;
    }
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solve(level, diff_arr, time_arr):
    
    ret = 0
    
    for i in range(1, len(diff_arr)):
        
        if diff_arr[i] <= level:
            ret += time_arr[i]
        else:
            diff = diff_arr[i] - level
            
            ret += time_arr[i - 1] * diff
            ret += time_arr[i] * (diff + 1)
    
    return ret


def solution(diffs, times, limit):
    
    n = len(diffs)
    
    diff_arr = [0] * (n + 1)
    time_arr = [0] * (n + 1)
    
    for i in range(n):
        diff_arr[i + 1] = diffs[i]
        time_arr[i + 1] = times[i]
    
    left = 1
    right = 100000
    ans = -1
    
    while left <= right:
        
        mid = (left + right) // 2
        
        result = solve(mid, diff_arr, time_arr)
        
        if result <= limit:
            ans = mid
            right = mid - 1
        else:
            left = mid + 1
    
    return ans